### PR TITLE
soundwire: intel: fix wrong register name in intel_shim_wake

### PR DESCRIPTION
**Pull request by libinyang**
When clearing the sdw wakests status, we should use SDW_SHIM_WAKESTS.

Signed-off-by: Libin Yang <libin.yang@intel.com>

### CODE REVIEW

**Review by bardliao — APPROVED** [review]
Nice catch @libinyang You may want to provide a fix tag

File: drivers/soundwire/intel.c [hunk 1]
```diff
@@ -449,7 +449,7 @@ static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
 		/* Clear wake status */
 		wake_sts = intel_readw(shim, SDW_SHIM_WAKESTS);
 		wake_sts |= (SDW_SHIM_WAKEEN_ENABLE << link_id);
-		intel_writew(shim, SDW_SHIM_WAKESTS_STATUS, wake_sts);
+		intel_writew(shim, SDW_SHIM_WAKESTS, (wake_sts & SDW_SHIM_WAKESTS_STATUS));
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: the register change is fine, I don't see why you added an AND operation that will kill what is done in line 451. Something's not right here.

**Review by plbossart — CHANGES_REQUESTED** [review]

**Comment by plbossart** [comment]
This was added in 4a17c441c7cb
````
4a17c441c7cb ("soundwire: intel: revisit SHIM programming sequences.")

````

**Comment by plbossart** [comment]
@libinyang I think this is the 'right' fix

````diff
diff --git a/drivers/soundwire/intel.c b/drivers/soundwire/intel.c
index 79ba0e3f6dac..9f532beb89bf 100644
--- a/drivers/soundwire/intel.c
+++ b/drivers/soundwire/intel.c
@@ -449,7 +449,7 @@ static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
                /* Clear wake status */
                wake_sts = intel_readw(shim, SDW_SHIM_WAKESTS);
                wake_sts |= (SDW_SHIM_WAKEEN_ENABLE << link_id);
-               intel_writew(shim, SDW_SHIM_WAKESTS_STATUS, wake_sts);
+               intel_writew(shim, SDW_SHIM_WAKESTS, wake_sts);
        }
        mutex_unlock(sdw->link_res->shim_lock);
 }
````

edit: I double-checked the hardware spec and BIT(0) needs to be cleared for link0, BIT(1) for link1, etc. So no you cannot do an AND with the status, you have set the BIT(link_id) and write back

**Comment by plbossart** [comment]
better fix

````diff
diff --git a/drivers/soundwire/intel.c b/drivers/soundwire/intel.c
index 79ba0e3f6dac..4b010d31506b 100644
--- a/drivers/soundwire/intel.c
+++ b/drivers/soundwire/intel.c
@@ -448,8 +448,8 @@ static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
 
                /* Clear wake status */
                wake_sts = intel_readw(shim, SDW_SHIM_WAKESTS);
-               wake_sts |= (SDW_SHIM_WAKEEN_ENABLE << link_id);
-               intel_writew(shim, SDW_SHIM_WAKESTS_STATUS, wake_sts);
+               wake_sts |= (SDW_SHIM_WAKEEN_STATUS << link_id);
+               intel_writew(shim, SDW_SHIM_WAKESTS, wake_sts);
        }
        mutex_unlock(sdw->link_res->shim_lock);
 }
````

File: drivers/soundwire/intel.c (hunk 1 above)
Comment (libinyang) [reply drivers/soundwire/intel.c:452]: @plbossart I referred the ACE1.x spec. It says: ``` (SNDWSC-1)*(1-SNDWMOD):0 : PREQ/WakeUp Status```  and  ```15:(SNDWSC-1)*(1-SNDWMOD)+1 : Reserved SW must use zeros for writes``` The cAVS2.x has the similar description. So I believe ```wake_sts & SDW_SHIM_WAKESTS_STATUS``` is the right writing. The only problem is how to set the ```SDW_SHIM_WAKESTS_STATUS```. For the ACE1.x and cAVS2.x, both are 0xf. And I'm still checking other platforms' setting (I don't have the spec so far). What about after I have checked other platforms setting, I submit another patch for ```SDW_SHIM_WAKESTS_STATUS``` setting?

**Comment by libinyang** [comment]
BTW: Based on my observation, it seems some other irq states are not cleared. I'm not sure it is correct or not. I'm still investigating on it.

File: drivers/soundwire/intel.c (hunk 1 above)
Comment (ranj063) [reply drivers/soundwire/intel.c]: @libinyang please don't refer embargoed codenames in public

**Review by RanderWang — DISMISSED** [review]
LGTM. Good catch!

**Comment by plbossart** [comment]
@libinyang can you please add a

````
Fixes: 4a17c441c7cb ("soundwire: intel: revisit SHIM programming sequences.")
````

**Review by RanderWang — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Review by plbossart — APPROVED** [review]
Nice work @libinyang, one of the worst bugs in a long time!

**Comment by plbossart** [comment]
@bardliao can we send this upstream asap, this should go as a fix for v5.17-rcX.

**Comment by bardliao** [comment]
> @bardliao can we send this upstream asap, this should go as a fix for v5.17-rcX.

sent

**Pull request merged**